src/crimson/osd/pg_map.cc

```diff
@@ -14,75 +14,140 @@ namespace crimson::osd {
 
 seastar::future<core_id_t> PGShardMapping::get_or_create_pg_mapping(
   spg_t pgid,
-  core_id_t core)
+  core_id_t core_expected)
 {
   LOG_PREFIX(PGShardMapping::get_or_create_pg_mapping);
   auto find_iter = pg_to_core.find(pgid);
   if (find_iter != pg_to_core.end()) {
-    ceph_assert_always(find_iter->second != NULL_CORE);
-    if (core != NULL_CORE) {
-      ceph_assert_always(find_iter->second == core);
+    auto core_found = find_iter->second;
+    assert(core_found != NULL_CORE);
+    if (core_expected != NULL_CORE && core_expected != core_found) {
+      ERROR("the mapping is inconsistent for pg {}: core {}, expected {}",
+            pgid, core_found, core_expected);
+      ceph_abort("The pg mapping is inconsistent!");
     }
-    return seastar::make_ready_future<core_id_t>(find_iter->second);
+    return seastar::make_ready_future<core_id_t>(core_found);
   } else {
-    return container().invoke_on(0,[pgid, core, FNAME]
-      (auto &primary_mapping) {
-      auto [insert_iter, inserted] = primary_mapping.pg_to_core.emplace(pgid, core);
-      ceph_assert_always(inserted);
-      ceph_assert_always(primary_mapping.core_to_num_pgs.size() > 0);
-      std::map<core_id_t, unsigned>::iterator core_iter;
-      if (core == NULL_CORE) {
-        core_iter = std::min_element(
-          primary_mapping.core_to_num_pgs.begin(),
-          primary_mapping.core_to_num_pgs.end(),
+    DEBUG("calling primary to add mapping for pg {} to the expected core {}",
+          pgid, core_expected);
+    return container().invoke_on(
+        0, [pgid, _core_expected=core_expected, FNAME](auto &primary_mapping) {
```

Comment:
[pgid, core_expected, ...] should just work?

Reply:
I wanted to update `core_expected` but this lambda doesn't worth to become mutable.

Refer to the following logic:
```c++
auto core_expected = _core_expected;
...
  core_expected = core_found;
...
assert(core_expected != NULL_CORE);
```

Probably I can rename to make this cleaner.

Reply:
Preserved `core_expected` and renamed to use `core_to_update` to store the updates.